Modules/_asynciomodule.c

```diff
@@ -1799,17 +1802,12 @@ future_new_iter(PyObject *fut)
     futureiterobject *it;
 
     asyncio_state *state = get_asyncio_state_by_def((PyObject *)fut);
-    if (!Future_Check(state, fut)) {
-        PyErr_BadInternalCall();
-        return NULL;
-    }
-
     ENSURE_FUTURE_ALIVE(state, fut)
 
-    if (fi_freelist_len) {
-        fi_freelist_len--;
-        it = fi_freelist;
-        fi_freelist = (futureiterobject*) it->future;
+    if (state->fi_freelist_len) {
+        state->fi_freelist_len--;
+        it = state->fi_freelist;
```

Comment:
Nit: remove the declaration in line 1802, and use intermingled declaration:

```suggestion
        futureiterobject *it = state->fi_freelist;
```

Reply:
It is also used in else block so won't work.